packages/aws-cdk/lib/util/work-graph.ts

```diff
@@ -0,0 +1,304 @@
+import { WorkNode, DeploymentState, StackNode, AssetBuildNode, AssetPublishNode } from './work-graph-types';
+
+export type Concurrency = number | Record<WorkNode['type'], number>;
+
+export class WorkGraph {
+  public readonly nodes: Record<string, WorkNode>;
+  private readonly readyPool: Array<WorkNode> = [];
+  private readonly lazyDependencies = new Map<string, string[]>();
+  public error?: Error;
+
+  public constructor(nodes: Record<string, WorkNode> = {}) {
+    this.nodes = { ...nodes };
+  }
+
+  public addNodes(...nodes: WorkNode[]) {
+    for (const node of nodes) {
+      const ld = this.lazyDependencies.get(node.id);
+      if (ld) {
+        for (const x of ld) {
+          node.dependencies.add(x);
+        }
+        this.lazyDependencies.delete(node.id);
+      }
+
+      this.nodes[node.id] = node;
+    }
+  }
+
+  public removeNode(nodeId: string | WorkNode) {
+    const id = typeof nodeId === 'string' ? nodeId : nodeId.id;
+    const removedNode = this.nodes[id];
+
+    this.lazyDependencies.delete(id);
+    delete this.nodes[id];
+
+    if (removedNode) {
+      for (const node of Object.values(this.nodes)) {
+        node.dependencies.delete(removedNode.id);
+      }
+    }
+  }
+
+  /**
+   * Return all nodes of a given type
+   */
+  public nodesOfType<T extends WorkNode['type']>(type: T): Extract<WorkNode, { type: T }>[] {
+    return Object.values(this.nodes).filter(n => n.type === type) as any;
+  }
+
+  /**
+   * Return all nodes that depend on a given node
+   */
+  public dependees(nodeId: string | WorkNode) {
+    const id = typeof nodeId === 'string' ? nodeId : nodeId.id;
+    return Object.values(this.nodes).filter(n => n.dependencies.has(id));
+  }
+
+  /**
+   * Add a dependency, that may come before or after the nodes involved
+   */
+  public addDependency(fromId: string, toId: string) {
+    const node = this.nodes[fromId];
+    if (node) {
+      node.dependencies.add(toId);
+      return;
+    }
+    let lazyDeps = this.lazyDependencies.get(fromId);
+    if (!lazyDeps) {
+      lazyDeps = [];
+      this.lazyDependencies.set(fromId, lazyDeps);
+    }
+    lazyDeps.push(toId);
+  }
+
+  public node(id: string) {
+    const ret = this.nodes[id];
+    if (!ret) {
+      throw new Error(`No node with id ${id} among ${Object.keys(this.nodes)}`);
+    }
+    return ret;
+  }
+
+  public absorb(graph: WorkGraph) {
+    this.addNodes(...Object.values(graph.nodes));
+  }
+
+  public hasFailed(): boolean {
```

Comment:
this used to be public because it was being used in deploy.ts but now it is only being used in the private `forAllArtifacts` method. I think we should make it private as well.

Reply:
I actually kinda like it. But you're right, we don't need it for now.